File: backend/app/schemas/sre.py
from typing import Literal

from pydantic import BaseModel, Field

ChaosMode = Literal["latency", "error", "timeout", "slow_query", "saturation"]
LoadScenario = Literal["steady", "checkout_rush", "search_storm"]


class ServiceOut(BaseModel):
    name: str
    tier: int
    owner: str
    endpoints: list[str]
    health: Literal["healthy", "degraded", "down"]
    version: str


class SignalPoint(BaseModel):
    ts: str
    rpm: float
    error_rate: float
    p95_ms: float


class SignalsOut(BaseModel):
    service: str
    window_minutes: int
    traffic_rpm: float
    error_rate: float
    latency_p50_ms: float
    latency_p95_ms: float
    latency_p99_ms: float
    saturation_pct: float
    synthetic: bool
    series: list[SignalPoint]


class SloOut(BaseModel):
    id: str
    service: str
    name: str
    kind: Literal["availability", "latency"]
    objective_pct: float
    window_days: int
    current_pct: float
    status: Literal["ok", "at_risk", "breached"]
    threshold_ms: float | None = None


class ErrorBudgetOut(BaseModel):
    slo_id: str
    objective: float
    achieved: float
    budget_remaining_pct: float
    burn_rate_1h: float
    burn_rate_6h: float
    status: Literal["ok", "at_risk", "breached"]


class ChaosToggleOut(BaseModel):
    target: str
    mode: ChaosMode
    magnitude: float
    active: bool
    expires_at: str | None = None


class ChaosToggleIn(BaseModel):
    target: str = Field(min_length=1, max_length=32)
    mode: ChaosMode
    magnitude: float = Field(default=500.0, ge=0, le=60000)
    ttl_seconds: int = Field(default=300, ge=1, le=3600)


class LoadRequest(BaseModel):
    scenario: LoadScenario = "steady"
    duration_seconds: int = Field(default=30, ge=1, le=600)
    rps: int = Field(default=5, ge=1, le=200)


class LoadResponse(BaseModel):
    status: str
    scenario: LoadScenario
    duration_seconds: int
    rps: int
    requests_planned: int

File: frontend/src/lib/sre.ts
import { api } from './api'

export interface SreService {
  name: string
  tier: number
  owner: string
  endpoints: string[]
  health: 'healthy' | 'degraded' | 'down'
  version: string
}

export interface SignalPoint {
  ts: string
  rpm: number
  error_rate: number
  p95_ms: number
}

export interface Signals {
  service: string
  window_minutes: number
  traffic_rpm: number
  error_rate: number
  latency_p50_ms: number
  latency_p95_ms: number
  latency_p99_ms: number
  saturation_pct: number
  synthetic: boolean
  series: SignalPoint[]
}

export interface Slo {
  id: string
  service: string
  name: string
  kind: 'availability' | 'latency'
  objective_pct: number
  window_days: number
  current_pct: number
  status: SloStatus
  threshold_ms: number | null
}

export type SloStatus = 'ok' | 'at_risk' | 'breached'

export interface ErrorBudget {
  slo_id: string
  objective: number
  achieved: number
  budget_remaining_pct: number
  burn_rate_1h: number
  burn_rate_6h: number
  status: SloStatus
}

export type ChaosMode = 'latency' | 'error' | 'timeout' | 'slow_query' | 'saturation'

export interface ChaosToggle {
  target: string
  mode: ChaosMode
  magnitude: number
  active: boolean
  expires_at: string | null
}

export type LoadScenario = 'steady' | 'checkout_rush' | 'search_storm'

export interface LoadResponse {
  status: string
  scenario: LoadScenario
  duration_seconds: number
  rps: number
  requests_planned: number
}

export const CHAOS_MODES: ChaosMode[] = ['latency', 'error', 'timeout', 'slow_query', 'saturation']
export const CHAOS_TARGETS = [
  'booking',
  'payments',
  'checkin',
  'flights',
  'notifications',
  'loyalty',
  'irrops',
]
export const LOAD_SCENARIOS: LoadScenario[] = ['steady', 'checkout_rush', 'search_storm']

export const MODE_UNITS: Record<ChaosMode, string> = {
  latency: 'ms delay',
  error: '% of requests failed',
  timeout: 'ms before timeout',
  slow_query: 'ms added to queries',
  saturation: '% worker capacity',
}

export const listServices = () => api<SreService[]>('/api/sre/services')

export const getSignals = (service: string, windowMinutes: number) =>
  api<Signals>(`/api/sre/services/${encodeURIComponent(service)}/signals?window_minutes=${windowMinutes}`)

export const listSlos = () => api<Slo[]>('/api/sre/slos')

export const getErrorBudget = (sloId: string) =>
  api<ErrorBudget>(`/api/sre/slos/${encodeURIComponent(sloId)}/error-budget`)

export const listChaos = () => api<ChaosToggle[]>('/api/sre/chaos')

export const armChaos = (body: {
  target: string
  mode: ChaosMode
  magnitude: number
  ttl_seconds: number
}) => api<ChaosToggle>('/api/sre/chaos', { method: 'POST', body: JSON.stringify(body) })

export const stopChaos = (target: string) =>
  api<{ status: string }>(`/api/sre/chaos/${encodeURIComponent(target)}`, { method: 'DELETE' })

export const startLoad = (body: {
  scenario: LoadScenario
  duration_seconds: number
  rps: number
}) => api<LoadResponse>('/api/sre/load', { method: 'POST', body: JSON.stringify(body) })

export function healthBadge(health: SreService['health']): string {
  return health === 'healthy' ? 'ok' : health === 'degraded' ? 'warn' : 'crit'
}

export function statusBadge(status: SloStatus): string {
  return status === 'ok' ? 'ok' : status === 'at_risk' ? 'warn' : 'crit'
}

export function clockTime(iso: string): string {
  return new Date(iso).toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' })
}

export function secondsUntil(iso: string | null): number {
  if (!iso) return 0
  return Math.max(0, Math.round((new Date(iso).getTime() - Date.now()) / 1000))
}
